### PR TITLE
Fix CGun::Fire server crash (fixes #6023)

### LINKED ISSUE #6023: Server crash when changing map on Linux x86-64 build
1. Start server
2. Connect to server with client
3. Change map to [Skill_Issue.zip](https://github.com/ddnet/ddnet/files/9970955/Skill_Issue.zip)
4. Linux x86-64 nightly build crashes, 16.5 too:
```
Thread 1 "DDNet-Server" received signal SIGSEGV, Segmentation fault.
0x00000000004c218d in ?? ()
(gdb) bt
#0  0x00000000004c218d in ?? ()
#1  0x00000000004ba7c1 in ?? ()
#2  0x00000000004eedd6 in ?? ()
#3  0x000000000040be14 in ?? ()
#4  0x00007ffff7c30290 in __libc_start_call_main (main=main@entry=0x409eb0, argc=argc@entry=1,
    argv=argv@entry=0x7fffffffe2d8) at ../sysdeps/nptl/libc_start_call_main.h:58
#5  0x00007ffff7c3034a in __libc_start_main_impl (main=0x409eb0, argc=1, argv=0x7fffffffe2d8,
    init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe2c8)
    at ../csu/libc-start.c:381
#6  0x000000000040c54a in ?? ()
```
With Asan or RelWithDebInfo (debug symbols without strip) no crash.

### PR DESCRIPTION
**Pull request by def-**
## Checklist

- [ ] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
Can you test whether that fixes the bug? You were able to reproduce it, weren't you?

File: src/game/server/entities/gun.cpp [hunk 1]
```diff
@@ -70,7 +70,11 @@ void CGun::Fire()
 	for(int i = 0; i < NumPlayersInRange; i++)
 	{
 		CCharacter *pTarget = apPlayersInRange[i];
-		const int &TargetTeam = pTarget->Team();
+		if(!pTarget)
```
Comment (def-) [inline src/game/server/entities/gun.cpp:73]: This fixed it

**Comment by def-** [comment]
It's not clear why this fixes the issue. We have the same behavior in other code too.

**Comment by def-** [comment]
I'll print-debug some more

**Comment by def-** [comment]
Printing out the pointer inside of FindEntities (`std::cout << Num << " " << pEnt << " " << ppEnts[Num] << std::endl;`) makes it impossible to reproduce this issue: `0 0x8d0460 0x8d0460`
Without this `cout` the value assigned to ppEnts[Num] is `0x0`.

Truly a Heisenbug, don't look too closely or it disappears.

**Comment by Learath2** [comment]
If it's reproducible, maybe gdb would work to see what's going on? Though I suspect some sort of optimization happening there will prevent you from peeking at the value of that pointer even in gdb.

**Comment by def-** [comment]
Can I set breakpoints without debug info? Enabling debug info makes the issue disappear too.

File: src/game/server/entities/gun.cpp [hunk 2]
```diff
@@ -70,7 +70,7 @@ void CGun::Fire()
 	for(int i = 0; i < NumPlayersInRange; i++)
 	{
 		CCharacter *pTarget = apPlayersInRange[i];
-		const int &TargetTeam = pTarget->Team();
+		const int TargetTeam = pTarget->Team();
```
Comment (heinrich5991) [inline src/game/server/entities/gun.cpp:73]: Is it maybe UB to create a reference to a temporary? The reference sounds very suspect to me.

**Comment by Learath2** [comment]
That sounds painful. Have you tried `-g` but with `-O2/3` maybe that'll keep the crash but it'll give a much worse debugging experience

File: src/game/server/entities/gun.cpp (hunk 2 above)
Comment (def-) [reply src/game/server/entities/gun.cpp:73]: Yes, I thought so too and thus I'm testing with all these weird references removed.

**Comment by def-** [comment]
I have a fear that this is a compiler bug in LTO, will try disabling that.

File: src/game/server/entities/gun.cpp (hunk 2 above)
Comment (heinrich5991) [reply src/game/server/entities/gun.cpp:73]: Can you cite a source for this? I find it weird that this is possible.

Comment (Learath2) [reply src/game/server/entities/gun.cpp:73]: C++17 15.2.5 -> 15.2.6p1 -> 7.4

The return of the function is non-reference, so we have a `prvalue` it gets materialized to an `xvalue` which is a `glvalue` thus 15.2.5 applies and the lifetime is extended.

**Comment by def-** [comment]
I have unsuccessfully tried to build a small reproducer.

With `CXXFLAGS=-fno-strict-aliasing` the problem disappears. I think we are doing undefined stuff with our casts and ubsan is not catching it.

**Comment by Learath2** [comment]
Now that I think about it, why is `CGun::Fire` even being called after a map change? I'll try to reproduce this myself as well, it's bizarre, I like it.

**Comment by def-** [comment]
My current theory is that in `CGun::Fire` `apPlayersInRange` is aliasing the `(CEntity **)apPlayersInRange` we pass to `FindEntities`. Thus the compiler's optimization ignores anything that happens inside of `FindEntities` and knows already that the value is still zero from the `mem_zero` previously.

**Comment by Learath2** [comment]
I couldn't figure out whether this is our fault or the compilers, but at least I'm fairly certain what is indeed happening is that gcc thinks that the cast breaks the aliasing. So, might as well get the fix in. Thanks

bors r+

**Pull request merged**